include/metall/detail/file.hpp

```diff
@@ -338,128 +251,124 @@ inline bool free_file_space([[maybe_unused]] const int fd,
     METALL_ERROR("{}", e.what());
     success = false;
   }
 
-  if (success) {
-    success &= metall::mtlldetail::fsync(destination_path);
-  }
-
+  mtlldetail::fsync(destination_path);
   return success;
 }
 
-#else
-
-inline bool copy_file_dense(const std::string &source_path,
-                            const std::string &destination_path) {
-  {
-    const ssize_t source_file_size = get_file_size(source_path);
-    const ssize_t actual_source_file_size = get_actual_file_size(source_path);
-    if (source_file_size == -1 || actual_source_file_size == -1) {
-      return false;
-    }
+#ifdef __linux__
+inline bool prepare_file_copy_linux(const std::filesystem::path &source_path,
+                                    const std::filesystem::path &destination_path,
+                                    int *src,
+                                    int *dst) {
+  *src = ::open(source_path.c_str(), O_RDONLY);
+  if (*src == -1) {
+    METALL_ERRNO_ERROR("Unable to open {}", source_path.c_str());
+    return false;
+  }
 
-    // If the source file is empty, just create the destination file and done.
-    if (source_file_size == 0 || actual_source_file_size == 0) {
-      create_file(destination_path);
-      return true;
-    }
+  struct stat st;
+  if (::fstat(*src, &st) == -1) {
+    METALL_ERRNO_ERROR("Unable to stat {}", source_path.c_str());
+    os_close(*src);
+    return false;
   }
 
-  {
-    std::ifstream source(source_path);
-    if (!source.is_open()) {
-      std::stringstream ss;
-      ss << "Cannot open: " << source_path;
-      METALL_ERROR(ss.str().c_str());
-      return false;
-    }
+  *dst = ::open(destination_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC);
+  if (*dst == -1) {
+    METALL_ERRNO_ERROR("Unable to open {}", destination_path.c_str());
+    os_close(*src);
+    return false;
+  }
 
-    std::ofstream destination(destination_path);
-    if (!destination.is_open()) {
-      std::stringstream ss;
-      ss << "Cannot open: " << destination_path;
-      METALL_ERROR(ss.str().c_str());
-      return false;
-    }
+  if (::fchmod(*dst, st.st_mode) == -1) {
+    METALL_ERRNO_ERROR("Unable to chmod {}", destination_path.c_str());
+    os_close(*src);
+    os_close(*dst);
+    return false;
+  }
 
-    destination << source.rdbuf();
-    if (!destination) {
-      std::stringstream ss;
-      ss << "Something happened in the ofstream: " << destination_path;
-      METALL_ERROR(ss.str().c_str());
-      return false;
-    }
+  return true;
+}
 
-    destination.close();
+inline bool copy_file_sparse_linux(int src, int dst) {
+  off64_t off = 0;
+  while ((off = ::lseek64(src, off, SEEK_DATA)) != -1) {
+    off64_t const hole_start = ::lseek64(src, off, SEEK_HOLE);
+    assert(hole_start != -1); // cannot fail, always an implicit hole at the end of the file
 
-    if (!metall::mtlldetail::fsync(destination_path)) {
+    if (::copy_file_range(src, &off, dst, NULL, hole_start - off, 0) == -1) {
```

Comment:
Not completely sure I got what's happening here. 

We copy things from src to dest and we are doing that between the holes. That means the resulting file will also have also so be sparse as well, right?

Reply:
We are copying just the data between the holes.
Data resides always between `start = fseek(SEEK_DATA)` .. `end = fseek(SEEK_HOLE)`.
The offset into dst is set to `NULL` which means "use the internal file offset", i.e. append at the current cursor. Which will be updated after each call.

```
The following semantics apply for off_in, and similar statements
apply to off_out:

If off_in is NULL, then bytes are read from fd_in starting
from the file offset, and the file offset is adjusted by the
number of bytes copied.
```
(See #Description https://man7.org/linux/man-pages/man2/copy_file_range.2.html)

Reply:
ok, that's what I concluded. sounds reasonable. Leafs me know wondering how exactly sparsity is implemented on file system and kernel level. but that's definitely out of scope here.